### PR TITLE
Added support for throwing a Java exception from the model compiler

### PR DESCRIPTION
**Pull request by JCash**
We've made the model compiler more robust and it will now catch any C/C++ errors, and then throw a Java error. This lets the editor handle such a case, and we minimise the risk of the user losing any edits in the project.

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: engine/modelc/src/java/com/dynamo/bob/pipeline/ModelImporter.java [hunk 1]
```diff
@@ -471,6 +472,18 @@ public static void main(String[] args) throws IOException {
             return;
         }
 
+        // TODO: Setup a java test suite for the model importer
+        int i = 0;
+        for (String arg : args)
```
Comment (Jhonnyg) [inline engine/modelc/src/java/com/dynamo/bob/pipeline/ModelImporter.java:477]: Why not use a for(int i=0; i < args.size(), i++) setup if you need the i in the body?

**Comment by JCash** [comment]
Linux:
```
Exception in thread "main" java.lang.RuntimeException: Exception in Defold JNI code. Signal 11
    0x7f638c059d73: /home/builder/engine/modelc/build/src/libmodelc_shared.so(+0x3d73) [0x7f638c059d73]
    0x7f638c059b5b: /home/builder/engine/modelc/build/src/libmodelc_shared.so(+0x3b5b) [0x7f638c059b5b]
    0x7f63b192f520: /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f63b192f520]
    0x7f63b1983a7c: /lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c) [0x7f63b1983a7c]
    0x7f63b192f476: /lib/x86_64-linux-gnu/libc.so.6(raise+0x16) [0x7f63b192f476]
    0x7f638c059cf9: /home/builder/engine/modelc/build/src/libmodelc_shared.so(+0x3cf9) [0x7f638c059cf9]
    0x7f638c078055: /home/builder/engine/modelc/build/src/libmodelc_shared.so(_Z32Java_ModelImporter_TestExceptionP7JNIEnv_P7_jclassP8_jstring+0x35) [0x7f638c078055]
    0x7f6394807a10: [0x7f6394807a10]
    # <- native
	at com.dynamo.bob.pipeline.ModelImporter.TestException(Native Method)
	at com.dynamo.bob.pipeline.ModelImporter.main(ModelImporter.java:481)
```

macos:
```
Exception in thread "main" java.lang.RuntimeException: Exception in Defold JNI code. Signal 11
    0x0000000106a51c10: (_ZN5dmJNIL23DefaultJniSignalHandlerEiPv+0x30)
    0x00007ff8023d8c1d: (_sigtramp+0x1d)
    0x00007ff80237622b: (__pthread_kill+0xb)
    0x00007ff8023adf7b: (pthread_kill+0x107)
    0x00007ff8022ba480: (raise+0x1a)
    0x0000000106a51dae: (_ZN5dmJNI20TestSignalFromStringEPKc+0x7e)
    0x0000000106a720ba: (_Z32Java_ModelImporter_TestExceptionP7JNIEnv_P7_jclassP8_jstring+0x3a)
    # <- native
	at com.dynamo.bob.pipeline.ModelImporter.TestException(Native Method)
	at com.dynamo.bob.pipeline.ModelImporter.main(ModelImporter.java:481)
```

Unfortunately, Win32 currently only gives me the addresses, nothing more.

File: engine/modelc/src/jni_util.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    // Caller has to free() the memory
+    char* GenerateCallstack();
+
+// for testing
+// Pass in a string containing any of the SIG* numbers
+    void TestSignalFromString(const char* signal);
+
+}
+
+#endif
```
Comment (Jhonnyg) [inline engine/modelc/src/jni_util.h:21]: newline

**Review by Jhonnyg — DISMISSED** [review]
👍

File: engine/modelc/src/java/com/dynamo/bob/pipeline/ModelImporter.java [hunk 3]
```diff
@@ -471,6 +472,18 @@ public static void main(String[] args) throws IOException {
...
+        {
+            if (arg.equalsIgnoreCase("--test-exception"))
+            {
+                ModelImporter.TestException("Testing exception: " + args[i+1]);
+                return; // exit code 0
+            }
+            ++i;
+        }
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/modelc/src/java/com/dynamo/bob/pipeline/ModelImporter.java:485]: good question, I have no good answer :)

**Pull request merged**